lib/compress/zstd_compress.c

```diff
@@ -3989,10 +3992,11 @@ size_t ZSTD_compressStream2( ZSTD_CCtx* cctx,
         FORWARD_IF_ERROR( ZSTD_initLocalDict(cctx) , ""); /* Init the local dict if present. */
         ZSTD_memset(&cctx->prefixDict, 0, sizeof(cctx->prefixDict));   /* single usage */
         assert(prefixDict.dict==NULL || cctx->cdict==NULL);    /* only one can be set */
+        if (cctx->cdict)
```

Comment:
After some more digging, the reason for the getting the correct clevel for filesizes + dictsizes that do not qualify for reloading is that downstream in `ZSTD_compressBegin_internal()`, the CCtx is readjusted using `ZSTD_resetCCtx_usingCDict()`. 

However, in the case of reloading, the function called was `ZSTD_resetCCtx_internal()` which doesn't take into account the `cdict`.

So I think the mitigation is to adjust the `params` upstream (which is a copy of requestedParams), and later on if we need to reload the dict and call `ZSTD_resetCCtx_internal()`, this very `params` which may have the `cdict` compression level, will be copied into the `cctx->appliedParams` during `ZSTD_resetCCtx_internal()`, and in this case, we don't modify the `cctx->requestedParams`.

I'm not sure if this is exactly what you meant by using the `appliedParams`, but it does seem like the `appliedParams` are mainly touched in the `ZSTD_resetCCtx_internal()`

Reply:
I think that's fine.

`appliedParams` is the place where the _final_ result of parameter adjustments will be stored.

Before reaching that point, `appliedParams` could also be used as a form of "temporary storage", where intermediate adjustment results are saved.
However, this could prove messy and confusing, hence separating the 2 topics, by creating and using a dedicated `params` structure for the purpose of temporary storage and intermediate adjustment is fine, probably clearer.